components/capi/src/custom_writeable.rs

```diff
@@ -0,0 +1,107 @@
+use std::{fmt, ptr};
+
+#[repr(C)]
+/// An object that can one can write UTF-8 strings to
+///
+/// This allows the C API to write to arbitrary kinds of objects, for example a
+/// C++ std::string or a char buffer.
+///
+/// [`icu4x_simple_writeable()`] can be used to write to a fixed-size char buffer.
+///
+/// May be extended in the future to support further invariants
+///
+/// ICU4XCustomWriteable will not perform any cleanup on `context` or `buf`, these are logically
+/// "borrows" from the FFI side.
+///
+/// # Safety invariants:
+///  - `flush()` and `grow()` will be passed `context` and it should always be safe to do so.
+///     `context` may be  null, however `flush()` and `grow()` must then be ready to receive it as such.
+///  - `buf` must be `cap` bytes long
+///  - `grow()` must either return null or a valid buffer of at least the requested buffer size
+///  - Rust code must call `ICU4XCustomWriteable::flush()` before releasing to C
+pub struct ICU4XCustomWriteable {
+    /// Context pointer passed to `grow()` and `flush()`. May be `null`.
+    ///
+    /// The pointer may reference structured data on the foreign side,
+    /// such as C++ std::string, used to reallocate buf.
+    context: *mut c_void,
+    /// The raw string buffer, which will be mutated on the Rust side.
+    buf: *mut u8,
+    /// The current filled size of the buffer
+    len: usize,
+    /// The current capacity of the buffer
+    cap: usize,
+    /// Called by Rust to indicate that there is no more data to write.
+    ///
+    /// Arguments:
+    /// - `context` (`*mut c_void`): The `context` field of this struct
+    /// - `length` (`usize`): The final length of the string in `buf`
+    flush: extern "C" fn(*mut c_void, usize),
+    /// Called by Rust to request more capacity in the buffer. The implementation should allocate a new
+    /// buffer and copy the contents of the old buffer into the new buffer.
+    ///
+    /// Arguments:
+    /// - `context` (`*mut c_void`): The `context` field of this struct.
+    /// - `capacity` (`*mut usize`): The requested capacity. Should be updated to reflect
+    ///    the actual capacity if the allocated buffer is larger than was requested.
+    ///
+    /// Returns: the newly allocated buffer, or `null` if allocation failed.
+    grow: extern "C" fn(*mut c_void, *mut usize) -> *mut u8,
+}
+
+impl ICU4XCustomWriteable {
+    /// Call this function before releasing the buffer to C
+    pub fn flush(&mut self) {
+        (self.flush)(self.context, self.len);
+    }
+}
+impl fmt::Write for ICU4XCustomWriteable {
+    fn write_str(&mut self, s: &str) -> Result<(), fmt::Error> {
+        let needed_len = self.len + s.len();
+        if needed_len > self.cap {
+            let mut new_cap = needed_len;
+            let newbuf = (self.grow)(self.context, &mut new_cap);
+            if newbuf.is_null() {
+                return Err(fmt::Error);
+            }
+            self.cap = new_cap;
+            self.buf = newbuf;
```

Comment:
question: I would have expected the implementation of grow to maintain these members, rather than updating them from the Rust side, since to me they are "owned" by the C side of the FFI. In that case, grow could simply return whether it succeeded or not. Is there a reason for not doing things this way?

Reply:
I look at these members as owned by the Rust side, the ownership model is cleaner that way. The foreign side _may_ have its own version of length and cap stored in `context`, but those only get updated by `grow` and `flush`

Reply:
I had a similar suggestion yesterday. If the field is for Rust, then I think it should not be exposed at all to the C side. Or, if it's being exposed, then all code that deals with it should be done in C. Right?

Reply:
It's an input. The entire struct is an _input_ to Rust from C, so C writes to it and Rust reads from it. There is no part of this struct that C reads from later, aside from the shared buffer pointer